### ISSUE TITLE
h5py/defs.c: No such file or directory

### ISSUE THREAD

**Issue by LanceWesterhoff** [h5py/h5py#537]
Hello-
I went through and tried to build h5py on both CentOS6 and CentOS5. I end up with the following error each time. Since it's a fresh download, I wonder if there is something odd about the way I am installing the package that the current version doesn't support (h5py-2.1.0 installed this same way without difficulty). I'm not seeing anything in the logs that would indicate the problem. One thought is that I'm installing h5py in a separate directory path (e.g. "tools") from my python install, but I think this would be a supported install. Any suggestions? Thanks!

```javascript

[lance@hostname h5py-2.4.0]$ uname -a
Linux hostname 2.6.32-431.5.1.el6.centos.plus.x86_64 #1 SMP Tue Feb 11 21:21:23 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
[lance@hostname h5py-2.4.0]$ cat /etc/redhat-release 
CentOS release 6.5 (Final)
[lance@hostname h5py-2.4.0]$ export LD_LIBRARY_PATH=/opt/python/Python-2.7.9-01262015/lib:${LD_LIBRARY_PATH}
[lance@hostname h5py-2.4.0]$ export LD_LIBRARY_PATH=/opt/python/Python-2.7.9-tools-01262015/lib:${LD_LIBRARY_PATH}
[lance@hostname h5py-2.4.0]$ export PATH=/opt/python/Python-2.7.9-01262015/bin:${PATH}
[lance@hostname h5py-2.4.0]$ export PATH=/opt/python/Python-2.7.9-tools-01262015/bin:$PATH
[lance@hostname h5py-2.4.0]$ export PYTHONPATH=/opt/python/Python-2.7.9-tools-01262015/lib/python2.7/site-packages
[lance@hostname h5py-2.4.0]$ gcc --version
gcc (GCC) 4.9.0
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

[lance@hostname h5py-2.4.0]$ /opt/python/Python-2.7.9-01262015/bin/python setup.py configure --hdf5=/opt/python/Python-2.7.9-tools-01262015/
running configure
Autodetected HDF5 1.8.14
********************************************************************************
                       Summary of the h5py configuration

    Path to HDF5: '/opt/python/Python-2.7.9-tools-01262015/'
    HDF5 Version: '1.8.14'
     MPI Enabled: False
Rebuild Required: True

********************************************************************************
[lance@hostname h5py-2.4.0]$ /opt/python/Python-2.7.9-01262015/bin/python setup.py install --prefix=/opt/python/Python-2.7.9-tools-01262015/
running install
running bdist_egg
running egg_info
writing requirements to h5py.egg-info/requires.txt
writing h5py.egg-info/PKG-INFO
writing top-level names to h5py.egg-info/top_level.txt
writing dependency_links to h5py.egg-info/dependency_links.txt
reading manifest file 'h5py.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'h5py.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
running build_ext
Autodetected HDF5 1.8.14
********************************************************************************
                       Summary of the h5py configuration

    Path to HDF5: '/opt/python/Python-2.7.9-tools-01262015/'
    HDF5 Version: '1.8.14'
     MPI Enabled: False
Rebuild Required: True

********************************************************************************
building 'h5py.defs' extension
gcc -pthread -fno-strict-aliasing -I/opt/python/Python-2.7.9-01262015/include -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DH5_USE_16_API -I/opt/python/Python-2.7.9-tools-01262015/include -I/opt/python/tmp2/h5py-2.4.0/lzf -I/opt/local/include -I/usr/local/include -I/opt/python/Python-2.7.9-tools-01262015/lib/python2.7/site-packages/numpy/core/include -I/opt/python/Python-2.7.9-01262015/include/python2.7 -c /opt/python/tmp2/h5py-2.4.0/h5py/defs.c -o build/temp.linux-x86_64-2.7/opt/python/tmp2/h5py-2.4.0/h5py/defs.o
gcc: error: /opt/python/tmp2/h5py-2.4.0/h5py/defs.c: No such file or directory
gcc: fatal error: no input files
compilation terminated.
error: command 'gcc' failed with exit status 1

```

**Comment by ulrikpedersen** [comment]
Hello,

I have come across the same issue and have tracked down the problem - but no idea how to fix it...

Basically the trouble starts when calling cythonize() in [setup_build.py L140](https://github.com/h5py/h5py/blob/master/setup_build.py#L140)

The cythonize function is meant to generate the C files from the h5py/*.pyx files - but fail (silently) to do so in my case. I have interrogated both input arg and return object from this function call and it appears  to be OK... I'm mystified - and not really familiar with cython.

I can work around it with an ugly hack: after the first failed setup.py build attempt, can use the cython shell tool to generate the C files: cython h5py/*.pyx - and when I then run the setup.py build again it works.

Hope that helps...

**Comment by LanceWesterhoff** [comment]
That worked ulrikpedersen  - thank you for your help!

Incidentally, if it will help in maintaining the package, we observe the same problem on a Mac. For posterity, this is what we added to our automated build script as a work around. I'm not sure that this issue should be marked as "resolved" however since there is obviously a problem.

```bash
if ( ! ${PYHOME}/${PYDEST}/bin/python setup.py build ) ; then
    echo "RUNNING cython"
    ${PYHOME}/${PYDEST}/bin/cython -v h5py/*.pyx
    echo "COMPLETD cython"
    ${PYHOME}/${PYDEST}/bin/python setup.py build
else
    echo "NOTE: CYTHON CALL WITHIN h5py BUILD WORKED!"
fi

```

**Comment by ulrikpedersen** [comment]
Glad my work-around works for you too @LanceWesterhoff! It really is a hack and doesn't fix the core issue that cythonize() doesn't do what it says on the tin... I'm still no closer to figuring this out - cython and setuptools is a very complicated business.

**Comment by andrewcollette** [comment]
@ulrikpedersen, @LanceWesterhoff, thank you for investigating this further.  We've seen this issue crop up a few times on the list but I'm flummoxed.

@LanceWesterhoff, do you happen to know if the Mac involved uses a case-sensitive file system?

**Comment by LanceWesterhoff** [comment]
The Mac filesystem is Journaled HFS+. I did not turn on case-sensitivity (HFS+ will generally preserve case but is not case sensitive).

The Linux boxes are of course case sensitive.

Hopefully that helps! We are likewise flummoxed. We really have no idea why this is having trouble. Unfortunately, it is unclear how to get any additional [helpful] verbosity.

**Comment by ulrikpedersen** [comment]
I looks like we're both using a python installation (and site packages) which are not installed in the usual 'standard' location on the (linux) system. On our site we build most of our python packages from source in order to install into our specialized environment. Of course we then set our environment path and pythonpath appropriately. Perhaps this issue is related to non-standard installation locations?

But for now it is weekend... :smile:

**Comment by LanceWesterhoff** [comment]
Here's our full build script. This has all of the env vars defined and may give you some insights. As you can see, we start out with a pretty clean environment. There's a little bit of editing I could do to further clear things up, but it seems effective. 

Have a good weekend - mine will kick in soon as well!

````bash
#!/bin/bash

set -e

date

export PATH=/usr/bin:/bin
unset MODULEPATH
unset LD_LIBRARY_PATH
unset MODULESHOME
unset QTLIB
unset QTDIR
unset QTINC
unset OE_LICENSE 
unset CHEMIXDIR
unset DYLD_LIBRARY_PATH
unset PYTHONPATH

echo "STARTING"

if [ -e /opt/gcc/current/bin ] ; then
    export PATH=/opt/gcc/current/bin:${PATH}
    export LD_LIBRARY_PATH=/opt/gcc/current/lib:${LD_LIBRARY_PATH}
elif [ -e /share/apps/gcc/${PLTFM_VER}/current/bin ] ; then
    export PATH=/share/apps/gcc/${PLTFM_VER}/current/bin:${PATH}
fi

which gcc
gcc --version
which g++
g++ --version

if [ "`hostname -s`" == "server" ] ; then
    PYHOME=/share/apps/python/${PLTFM_VER}
else
    PYHOME=/opt/python
fi

PYVER=Python-2.7.9
PYDEST=${PYVER}-`date +%m%d%C%y`
TOOLDEST=${PYVER}-tools-`date +%m%d%C%y`

HDFVER=hdf5-1.8.14
HDFPYVER=h5py-2.4.0
SCIPYVER=scipy-0.15.1
NUMPYVER=numpy-1.9.1
CYTHVER=Cython-0.21.2

export PATH=${PYHOME}/${PYDEST}/bin:${PATH}
export LD_LIBRARY_PATH=${PYHOME}/${PYDEST}/lib:${LD_LIBRARY_PATH}
export DYLD_LIBRARY_PATH=${PYHOME}/${PYDEST}/lib:${DYLD_LIBRARY_PATH}
export CFLAGS="-I${PYHOME}/${PYDEST}/include ${CFLAGS}"
export CPPFLAGS="-I${PYHOME}/${PYDEST}/include ${CPPFLAGS}"
export CXXFLAGS="-I${PYHOME}/${PYDEST}/include ${CXXFLAGS}"
export LDFLAGS="-L${PYHOME}/${PYDEST}/lib ${LDFLAGS}"
export PYTHONPATH=${PYHOME}/${PYDEST}/lib/python2.7/site-packages

echo "Installing .... $PYDEST in ${PYHOME}"

WORKDIR=$PWD

if [ "$1" == "all" ] ; then

    if [ -e ${PYHOME}/${PYDEST}/bin ] ; then
        rm -rf ${PYHOME}/${PYDEST}/
    fi
    if [ -e ${PYHOME}/${TOOLDEST}/bin ] ; then
        rm -rf ${PYHOME}/${TOOLDEST}/
    fi

    cp /share/network/BUILD/TarBalls/${PYVER}.tar.gz .
    cp /share/network/BUILD/TarBalls/setuptools-0.6c11.tar.gz .
    cp /share/network/BUILD/TarBalls/docutils-0.10.tar.gz .
    cp /share/network/BUILD/TarBalls/zlib-1.2.7.tar.gz .
    cp /share/network/BUILD/TarBalls/${HDFVER}.tar.gz .
    cp /share/network/BUILD/TarBalls/${NUMPYVER}.tar.gz .
    cp /share/network/BUILD/TarBalls/${HDFPYVER}.tar.gz .
    cp /share/network/BUILD/TarBalls/${CYTHVER}.tar.gz .

    tar zxvf ${PYVER}.tar.gz
    mkdir -p ${PYHOME} ; cd ${PYHOME} ; rm -rf ${PYDEST}
    rm -rf ${WORKDIR}/build ; mkdir ${WORKDIR}/build
    cd ${WORKDIR}/build

    ../${PYVER}/configure --prefix=${PYHOME}/${PYDEST} --with-ssl --enable-shared
    make -j8
    make install

    which python

    cd ${WORKDIR}
    rm -rf ${WORKDIR}/setuptools-0.6c11 ; tar zxvf setuptools-0.6c11.tar.gz
    cd ${WORKDIR}/setuptools-0.6c11
    ${PYHOME}/${PYDEST}/bin/python setup.py build
    ${PYHOME}/${PYDEST}/bin/python setup.py install --prefix=${PYHOME}/${PYDEST}

    cd ${WORKDIR}
    rm -rf ${WORKDIR}/docutils-0.10 ; tar zxvf docutils-0.10.tar.gz
    cd ${WORKDIR}/docutils-0.10
    ${PYHOME}/${PYDEST}/bin/python setup.py build
    ${PYHOME}/${PYDEST}/bin/python setup.py install --prefix=${PYHOME}/${PYDEST}

    cd ${WORKDIR}
    rm -rf ${WORKDIR}/${CYTHVER} ; tar zxvf ${CYTHVER}.tar.gz
    cd ${WORKDIR}/${CYTHVER}
    ${PYHOME}/${PYDEST}/bin/python setup.py build
    ${PYHOME}/${PYDEST}/bin/python setup.py install --prefix=${PYHOME}/${PYDEST}

fi

# build the support libs version as well
export PATH=${PYHOME}/${TOOLDEST}/bin:${PATH}
export LD_LIBRARY_PATH=${PYHOME}/${TOOLDEST}/lib:${LD_LIBRARY_PATH}
export DYLD_LIBRARY_PATH=${PYHOME}/${TOOLDEST}/lib:${DYLD_LIBRARY_PATH}
export CFLAGS="-I${PYHOME}/${TOOLDEST}/include ${CFLAGS}"
export CPPFLAGS="-I${PYHOME}/${TOOLDEST}/include ${CPPFLAGS}"
export CXXFLAGS="-I${PYHOME}/${TOOLDEST}/include ${CXXFLAGS}"
export LDFLAGS="-L${PYHOME}/${TOOLDEST}/lib ${LDFLAGS}"
export PYTHONPATH=${PYHOME}/${TOOLDEST}/lib/python2.7/site-packages:${PYTHONPATH}
export HDF5_PATH=${PYHOME}/${TOOLDEST}

cd ${WORKDIR}
if [ "$1" == "all" ] ; then
    rm -rf zlib-1.2.7 ; gunzip zlib-1.2.7.tar.gz ; tar xvf zlib-1.2.7.tar
    cd ${WORKDIR}/zlib-1.2.7
    ./configure --prefix=${PYHOME}/${TOOLDEST}
    make -j8
    make install

    cd ${WORKDIR}
    rm -rf ${HDFVER} ; tar zxvf ${HDFVER}.tar.gz
    cd ${WORKDIR}/${HDFVER}
    ./configure --prefix=${HDF5_PATH}
    make -j8
    make install

    cd ${WORKDIR}
    rm -rf ${WORKDIR}/${NUMPYVER} ; tar zxvf ${NUMPYVER}.tar.gz
    cd ${WORKDIR}/${NUMPYVER}
    BLAS=None LAPACK=None ATLAS=None ${PYHOME}/${PYDEST}/bin/python setup.py build
    ${PYHOME}/${PYDEST}/bin/python setup.py install --prefix=${PYHOME}/${TOOLDEST}
fi

cd ${WORKDIR}
rm -rf ${WORKDIR}/${HDFPYVER} ; tar zxvf ${HDFPYVER}.tar.gz
cd ${WORKDIR}/${HDFPYVER}
${PYHOME}/${PYDEST}/bin/python setup.py configure --hdf5=$HDF5_PATH --hdf5-version=1.8.14
if ( ! ${PYHOME}/${PYDEST}/bin/python setup.py build ) ; then
    echo "RUNNING cython"
    ${PYHOME}/${PYDEST}/bin/cython -v h5py/*.pyx
    echo "COMPLETD cython"
    ${PYHOME}/${PYDEST}/bin/python setup.py build
else
    echo "NOTE: CYTHON CALL WITHIN h5py BUILD WORKED!"
fi
${PYHOME}/${PYDEST}/bin/python setup.py install --prefix=${PYHOME}/${TOOLDEST}

echo "FINISHED"

date



````

**Comment by andrewcollette** [comment]
Possible explanation, to do with older versions of setuptools:

https://groups.google.com/d/msg/h5py/SF44bd5pjoo/-fldvc-F4hYJ

**Comment by ulrikpedersen** [comment]
Yes, that does look like the right explanation. I have also done a bit of research and found the same explanation regarding an old version of setuptools - I am using 0.6.11. I tracked some ticket (and promptly lost the link in the long history of searches, apologies) which seem to indicate that this was fixed in Distribute which was then merged with setuptools at version 0.7.

Then I updated my setuptools to the latest version 12.0.5 and this did indeed fix the problem - .c files are now generated from the .pyx files. Great :smiley: 

Updating setuptools however, can be quite tricky and not something that I will embark on for production immediately... An idea is to do what numpy (and maybe scipy) does which seems a bit of a hack at first: if the .c files are not generated, simply run the system 'cython' command (subprocess). Numpy does that as part of the setup.py with this [cythonize.py script](https://github.com/numpy/numpy/blob/master/tools/cythonize.py)

That way the build would function regardless even for users with old setuptools (<0.7). I have not yet had time to try this out.
